Content.Server/Automod/AutomodManager.cs

```diff
@@ -0,0 +1,260 @@
+                _log.Error($"AutomodActionGroupPrototype \"{filter.ActionGroup}\" not found.");
+                continue;
+            }
+
+            var skip = false;
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                if (!censorAction.Skip(inputText, textMatches))
+                    continue;
+
+                skip = true;
+                break;
+            }
+            if (skip)
+                continue;
+
+            var displayName = GetDisplayName(filter);
+
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                blocked &= censorAction.RunAction(session, inputText, textMatches, filter, displayName, _entMan);
+            }
+        }
+
+        return blocked;
+    }
+
+    private string GetDisplayName(AutomodFilterDef filter)
+    {
+        return filter.Id != null ? $"{filter.DisplayName} (automod#{filter.Id})" : $"{filter.DisplayName} (automod#na)";
+    }
+
+    #endregion
+
+    #region Reload
+
+    public async void ReloadAutomodFilters()
+    {
+        _regexFilters.Clear();
+
+        var filters = await _db.GetAllAutomodFiltersAsync();
+
+        foreach (var filter in filters)
+        {
+            AddFilter(filter);
+        }
+    }
+
+    #endregion
+
+    #region Create
+
+    public async void CreateFilter(string pattern,
+        AutomodFilterType filterType,
+        string actionGroup,
+        AutomodTarget targets,
+        string name)
+    {
+        CreateFilter(new AutomodFilterDef(pattern, filterType, actionGroup, targets, name));
+    }
+
+    public async void CreateFilter(AutomodFilterDef automod)
+    {
+        automod = await _db.AddAutomodFilterAsync(automod);
+        AddFilter(automod);
+    }
+
+    private void AddFilter(AutomodFilterDef automod)
+    {
+        foreach (AutomodTarget targetFlag in Enum.GetValues(typeof(AutomodTarget)))
+        {
+            if (targetFlag == AutomodTarget.None)
+                continue;
+            if (!automod.TargetFlags.HasFlag(targetFlag))
+                continue;
+
+            if (automod.FilterType == AutomodFilterType.Regex)
+            {
+                if (!_regexFilters.TryGetValue(targetFlag, out var list))
+                {
+                    list = new Dictionary<Regex, AutomodFilterDef>();
+                    _regexFilters.Add(targetFlag, list);
+                }
+
+                list.Add(new Regex(automod.Pattern), automod);
+            }
+            // TODO other filter types
+        }
+    }
+
+    #endregion
+
+    #region Get
+
+    public async Task<AutomodFilterDef?> GetFilter(int id)
+    {
+        return await _db.GetAutomodFilterAsync(id);
+    }
+
+    #endregion
+
+    #region Edit
+
+    public void EditFilter(AutomodFilterDef automodFilterDef)
+    {
+        _db.EditAutomodFilterAsync(automodFilterDef);
+    }
+
+    #endregion
+
+    #region Remove
+
+    public async Task<bool> RemoveFilter(int id)
```

Comment:
Most of this code is duplicated with the below function

Reply:
Consolidated in [dc06e4b](https://github.com/space-wizards/space-station-14/pull/28570/commits/dc06e4b46d4fc594290ccfd3db2936b9b32590cc)